docs/proposals/14-wireguard-based-cluster-vpn.md

```diff
@@ -0,0 +1,178 @@
+
+The WireGuard endpoint on the seed is exposed via a load balancer. We propose to use kubelink to manage the WireGuard configuration/device on the seed. We consider the management of the WireGuard endpoint to be complex especially in error situations which is the reason for utilizing kubelink as there is already significant experience managing an endpoint.
+
+On the shoot, we create the keys and acquire the WireGuard IP in the standard secret generation. The data is added as a secret to the control plane and to the shootstate. The vpn shoot deployment is extended to include the WireGuard device setup inside the vpn shoot pod network. For certain infrastructures (AWS), we need a re-advertiser to resolve the seed WireGuard endpoint and evaluate whether the IP address changed.
+
+The load balancer exposing the open vpn endpoint is no longer required and therefore removed if this functionality is used.
+
+As we want to slowly expand the usage of the WireGuard solution, we propose to introduce a feature gate for it. Furthermore, since the WireGuard network requires a separate network range, we propose to introduce a new section to the seed settings with two additional flags (enabled & cidr):
+
+```
+apiVersion: core.gardener.cloud/v1beta1
+kind: Seed
+metadata:
+  name: my-seed
+  ...
+spec:
+  ...
+  settings:
+  ...
+    wireguard:
+      enabled: true
+      cidr: 192.168.0.0/22
+```
+
+Last but not least, we propose to introduce an annotation to the shoots to enable/disable the WireGuard tunnel explicitly.
+
+```
+apiVersion: core.gardener.cloud/v1beta1
+kind: Shoot
+metadata:
+  name: my-shoot
+  annotations:
+    alpha.featuregates.shoot.gardener.cloud/wireguard-tunnel: "true"
+  ...
+```
+
+Using this approach, it is easy to switch the solution on and off, i.e. migrate the shoot clusters automatically during ordinary reconciliation.
+
+## High Availability
+
+There is an issue if the node that hosts the WireGuard endpoint fails. The endpoint is migrated to another node however the time required to do this might exceed the budget for downtimes although one could argue that a disruption of less than 30 seconds to 1 minute does not qualify as a downtime and will in almost all cases not noticeable by end users. 
+
+In this case we also assume that TCP connections won't be interrupted - they would just appear to hang. We will confirm this as part of the development and testing effort.
+
+As a mitigation we propose to instantiate 2 Kubelink instances in the seed cluster. The instances must run on different nodes (if possible but we assume a proper seed cluster has more than one node). Each shoot cluster connects to both endpoints. This means that the OpenVPN server is reachable with two different IP addresses. The VPN seed sidecars must attempt to connect to both of them and will continue to do so. The "Persistent Keepalive" feature is set to 21 seconds by default but could be reduced. Due to the redundancy this however appears not to be necessary.
```

Comment:
> (if possible but we assume a proper seed cluster has more than one node).

Not sure, I guess there might be use-cases were this assumption does not apply (e.g., Gardener on Bare Metal with large physical machines).

Reply:
If we just have one node we have a single point of failure anyway. Looking at the core networking interfaces (and WireGuard is one of them), if one has an issue possibly all with have an issue in which case a dysfunctional WireGuard is probably not a real problem if network down completely.

I would argue that the proposed solution does not add a failure domain if the seed cluster consists of one large bare metal host.

Reply:
Yeah, probably true, I just wanted to add this aspect for completeness.